templates/search/search.html

```diff
@@ -201,13 +242,157 @@
+                stopSound();
+            })
+            .click(function() {
+                stopSound();
+                playAudio($(this).parent().parent(), false);
+            });
+    }
+
+    var clustering_trial_number = 0;
+    // function for requesting clustering
+    function request_clustering()
+    {
+        $.get({% url 'clustering-facet' %} + '?{{ url_query_params_string | safe }}', 
+                {}, 
+                function( data ) {
+                    clustering_trial_number += 1;
+                    if (data.status === 'pending') {
+                        if (clustering_trial_number < 1000) {
+                            setTimeout(() => {
+                                request_clustering();
+                            }, 500);
+                        } else {
+                            $('#facet-loader').hide();
+                            $('#cluster-fail-icon').show();
+                        }
+                    } else if (data.status === 'failed') {
+                        // clustering failed
+                        $('#facet-loader').hide();
+                        $('#cluster-fail-icon').show();
+                    } else {
+                        $('#facet-loader').hide();
+                        $('#show-clustering-graph-button').show();
+                        $('#cluster-labels').html(data);
+                        $('#clusters-div').replaceWith(data);
+                        // add cluster colors
+                        var num_clusters = Math.max(
+                            ...Array.from($('.clustering-facet').map(
+                                (e, l)=>parseInt($(l).attr('cluster-id')))
+                            )
+                        ) + 1;
+                        $('#cluster-labels').find('.clustering-facet').each((i, l) => {
+                            $(l).find('a').css("color", cluster2color($(l).attr('cluster-id'), num_clusters));
+                            $(l).find('a').css('font-weight', 'bold');
+                        });
+                        enableAudioClusterExamples();
+
+                        $('.cluster-link-button').click(function () {
+                            $("#clustering-graph-modal").show();
+                            var clusterId = $(this).attr('cluster-id');
+                            showCluster = clusterId;
+                            if (graph !== undefined && graphLoaded === false) {
+                                activateGraph(graph, clusterId);
+                                graphLoaded = true;
+                            }
+                        });
+
+                        $('#close-modal-button').click(function () {
+                            closeGraphModal();
+                        })
+
+                        // request clustered graph
+                        // set graph as a global variable
+                        $.get("{% url 'clustered-graph-json' %}"+"?{{url_query_params_string | safe}}", {
```

Comment:
Here, I use `| safe` because I need to have the url unescaped. 
Did not find another way to do that, but it may not be recommended to do that.